File: solutions/search-in-a-binary-search-tree/search-in-a-binary-search-tree.cpp
// @Title: Search in a Binary Search Tree
// @Author: colinjxc
// @Date: 2022-01-27T17:15:14+08:00
// @URL: https://leetcode-cn.com/problems/search-in-a-binary-search-tree


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* searchBST(TreeNode* root, int val) {
        queue<TreeNode*> _q;
        _q.push(root);

        TreeNode* node = nullptr;
        while (!_q.empty()) {
            node = _q.front();
            _q.pop();
            if (node == nullptr || node->val == val) {
                return node;
            } else if (node->val < val) {
                _q.push(node->right);
            } else {
                _q.push(node->left);
            }
        }

        return nullptr;
    }
};

File: solutions/search-in-a-binary-search-tree/search-in-a-binary-search-tree.go
// @Title: Search in a Binary Search Tree
// @Author: colinjxc
// @Date: 2021-05-23T18:07:33+08:00
// @URL: https://leetcode-cn.com/problems/search-in-a-binary-search-tree


/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func searchBST(root *TreeNode, val int) *TreeNode {
	queue := list.New()
	node := queue.PushBack(root).Value.(*TreeNode)

	for queue.Len() != 0 {
		node = queue.Remove(queue.Front()).(*TreeNode)
		if node == nil || node.Val == val {
			break
		} else if val > node.Val {
			queue.PushBack(node.Right)
		} else {
			queue.PushBack(node.Left)
		}
	}
	return node
}
